capstone/src/test/java/com/google/step/clients/MetricsClientTests.java

```diff
@@ -0,0 +1,351 @@
+        //          "week": current week number,
+        //          "count": 100
+        //      }
+
+        // Using calendar for this test is necessary since the method in the
+        // metricsClient uses the calendar as well to get the current date.
+        Calendar calendar = Calendar.getInstance();
+        int year = calendar.get(Calendar.YEAR);
+        int week = calendar.get(Calendar.WEEK_OF_YEAR);
+
+        Entity entity1 = new Entity("PageViews");
+        entity1.setProperty("restaurantKey", "1");
+        entity1.setProperty("restaurantName", "Marlows");
+        entity1.setProperty("year", year);
+        entity1.setProperty("week", week);
+        entity1.setProperty("count", 100);
+
+        Entity entity2 = new Entity("PageViews");
+        entity2.setProperty("restaurantKey", "1");
+        entity2.setProperty("restaurantName", "Marlows");
+        entity2.setProperty("year", 2015);
+        entity2.setProperty("week", 7);
+        entity2.setProperty("count", 200);
+
+        datastoreService.put(entity1);
+        datastoreService.put(entity2);
+
+        Map<String, Object> resultMap = metricsClient.getCurrentPageViews("Marlows", "1");
+        Assert.assertEquals("Year", year, resultMap.get("year"));
+        Assert.assertEquals("Week", week, resultMap.get("week"));
+        Assert.assertEquals("Count", "100", resultMap.get("count").toString());
+    }
+    @Test
+    public void testGetCurrentPageViewsWithoutExistingEntity() {
+        // Tests if the method correctly created an entity and puts into dataStore
+        // if no entity exists for this week currently.
+        // Result:
+        //      {
+        //          "year": current year (2020),
+        //          "week": current week number,
+        //          "count": 0
+        //      }
+
+        // Using calendar for this test is necessary since the method in the
+        // metricsClient uses the calendar as well to get the current date.
+        Calendar calendar = Calendar.getInstance();
+        int year = calendar.get(Calendar.YEAR);
+        int week = calendar.get(Calendar.WEEK_OF_YEAR);
+
+        Map<String, Object> resultMap = metricsClient.getCurrentPageViews("Marlows", "1");
+        Assert.assertEquals("Year", year, resultMap.get("year"));
+        Assert.assertEquals("Week", week, resultMap.get("week"));
+        Assert.assertEquals("Count", "0", resultMap.get("count").toString());
+        Assert.assertEquals(1, datastoreService.prepare(new Query("PageViews")).countEntities(withLimit(10)));
+    }
+
+    @Test
+    public void testCheckBeginningDate() {
+        // Checks the map is being updated correctly when the year and week have only been initialized.
+        // Result: {"year": 1, "week": 1}
+        Map<String, Integer> map = new HashMap<>();
+        map.put("year", Integer.MAX_VALUE);
+        map.put("week", Integer.MAX_VALUE);
+        metricsClient.checkEarlierDate(map, 1, 1);
```

Comment:
I assume that this is the reason that `checkEarlierDate` is protected. I think that you should be designing the main code as if that code doesn't know the tests exist. Don't get me wrong; that isn't to say that you shouldn't design your code to be testable. What I mean to say is that you shouldn't have to change your access modifiers for the explicit purpose of testing. Let things like that represent how you think the code should be used by others, not by the unit tests. Tests should only really have access to everything that an external user should have access to, as that's what others will be using.

I see where you're coming from with testing the checkEarlierDate functionality. What I would suggest instead is checking this functionality through the public facing methods and the Datastore helper, similar to how you did it in some other tests.

Reply:
Ahhh yeah that makes a lot of sense. Thanks for the advice! It has been changed so the tests only actually test for the public facing methods.